default/AI/ShipDesignAI.py

```diff
@@ -1308,6 +1256,28 @@ def _calc_rating_for_name(self):
         self.update_stats(ignore_species=True)
         return self.structure
 
+    def _adjusted_production_cost(self):
+        """Return a production cost adjusted by the number of ships we have.
+
+        Building one warship of rating 2R and cost 2C is effectively cheaper than building 2 warships of rating R and
+        cost C due to fleet upkeep considerations even if they have the same raw rating/cost ratio.
+        The significance of this fleet upkeep efficiency increases with the raw fleet upkeep rate, with the expected
+        lifetime of the ship under consideration (i.e., with a longer expected lifespan it will affect the future
+        construction of a greater number of ships), and it also increases with the number of ships of this design
+        expected to be created. In the calculation below, the empire's current total shipcount is taken as a rough proxy
+        for both of the duration and extent factors.
+        Note: This same sort of adjustment would be valid for troop ships, could theoretically have some applicability
+        to scout ships since there could conceivably be a tradeoff for more scout ships of lower rating/range, but
+        would really not be applicable to colony ships since in that case there is really not a potential tradeoff
+        between number of ships and capacity.
+
+        :return: adjusted production cost
+        :rtype: float
+        """
+        # TODO: Consider total pp production output as additional factor
```

Comment:
I have that on my TODO list for quite some time now but am never too happy with the places where it could be.
First problem is caching - if PP output changes, then the cost function isn't cached anymore. Caching PP output though is rather pointless as it will change almost on a turn-by-turn basis. The same problem holds in principle for the new adjusted cost, too.

I am currently leaning towards making an initial pass as it happens right now and then if the production cost of the prefered design is too high, make a second pass with some flag which alters the evaluation code, does not cache results and alters the filters to allow cheaper parts.

However, this would still require significant changes to the "filling" algorithm itself as the current algorithm does not handle being on the upper limit of something very well / at all.

An alternative within the current framework is to split military ships into some "big" ships and "small" ships where we enforce the latter to be cheap without reference to the current pp output (e.g. we could divide by production cost squared). The production code then may decide which one we want to build on the pp-output basis.